runners/portability/java/build.gradle

```diff
@@ -108,95 +120,113 @@ startLocalJobService.finalizedBy stopLocalJobService
 
 /**
  * Runs Java ValidatesRunner tests against the Universal Local Runner (ULR) aka local_job_service_main
- * with subprocess SDK harness environments.
+ * with the specified environment type.
  */
-task ulrValidatesRunnerTests(type: Test) {
-  dependsOn ":sdks:java:container:docker"
-
-  if (!project.hasProperty("localJobServicePortFile")) {
-    dependsOn startLocalJobService
+def createUlrValidatesRunnerTask = { name, environmentType ->
+  Task vrTask = tasks.create(name: name, type: Test, group: "Verification") {
+    description "PortableRunner Java docker ValidatesRunner suite"
+    classpath = configurations.validatesRunner
+    systemProperty "beamTestPipelineOptions", JsonOutput.toJson([
+        "--runner=TestUniversalRunner",
+        "--experiments=beam_fn_api",
+        "--defaultEnvironmentType=${environmentType}",
+        "--localJobServicePortFile=${localJobServicePortFile}"
+    ])
+    testClassesDirs = files(project(":sdks:java:core").sourceSets.test.output.classesDirs)
+    useJUnit {
+      includeCategories 'org.apache.beam.sdk.testing.ValidatesRunner'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesGaugeMetrics'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesOnWindowExpiration'
+      excludeCategories 'org.apache.beam.sdk.testing.UsesBundleFinalizer'
```

Comment:
`UsesBundleFinalizer` is also new since the initial setup of the VR gradle task.

Reply:
Do we need a TODO for any of these exclusions?

Reply:
I don't believe so. I think all excluded categories are essentially feature enhancements for the roadmap. For individual test cases that do not correspond to a category they each have a Jira, since we should aim to support features entirely or not at all. We should also avoid adding fine-grained categories if they do not correspond to real general concepts.